packages/gatsby/src/query/query-compiler.js

```diff
@@ -443,6 +444,10 @@ const determineUsedFragmentsForDefinition = (
       [Kind.FRAGMENT_SPREAD]: node => {
         const name = node.name.value
         const fragmentDefinition = definitionsByName.get(name)
+        if (visitedFragmentDefinitions.has(fragmentDefinition)) {
+          return
+        }
+        visitedFragmentDefinitions.add(fragmentDefinition)
         if (fragmentDefinition) {
```

Comment:
This implies that you may be adding primitives to the `visitedFragmentDefinitions` set.

If nothing else, I would bet that we could move the 447-450 to be inside the falsy check to prevent falsy values being added to the queue. Or if the value is never falsy, drop the check.

Also, if `name` is a unique descriptor of these objects then maybe that would be as good a value to use for this purpose.

Reply:
Right, `fragmentDefinition` can be `undefined` in an edge case. Will move the check inside `if (fragmentDefinition) { }`